File: Cpp/K-th Symbol in Grammar.cpp
// 0ms, 100.00%; 8.5MB, 8.33%
class Solution
{
public:
    int kthGrammar(int N, int K)
    {
        bool isNot = false;
        while (K > 2) {
            int row = ceil(log2(K));
            if (row & 1) {
                isNot ^= true;
                K -= (1 << (row - 1));
            } else {
                K = (1 << row) - K + 1;
            }
        }
        return (isNot ? 2 - K : K - 1);
    }
};

File: Python3/K-th Symbol in Grammar.py
# 36ms, 64.00%; 13.9MB, 25.00%
from math import ceil, log2

class Solution:
    def kthGrammar(self, _, K: int) -> int:
        is_not = False
        while K > 2:
            row = int(ceil(log2(K)))
            if row & 1:
                is_not ^= True
                K -= 2 ** (row - 1)
            else:
                K = 2 ** row - K + 1
        return 2 - K if is_not else K - 1
